## Exploring Hacker News Posts

Hacker News is a site started by the startup incubator Y Combinator, where user-submitted stories (known as "posts") are voted and commented upon, similar to reddit. Hacker News is extremely popular in technology and startup circles, and posts that make it to the top of Hacker News' listings can get hundreds of thousands of visitors as a result.

In this project, we're specifically interested in posts whose titles begin with either Ask HN or Show HN. Users submit Ask HN posts to ask the Hacker News community a specific question. Users submit Show HN posts to show the Hacker News community a project, product, or just generally something interesting.

We'll compare these two types of posts to determine the following:

- Do Ask HN or Show HN receive more comments on average?
- Do posts created at a certain time receive more comments on average?

## Opening Data

We will use data in [here](https://www.kaggle.com/hacker-news/hacker-news-posts).

In [3]:
from csv import reader
opened_file=open('hacker_news.csv')
read_file=reader(opened_file)
hn=list(read_file)
hn[:5]

[['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at'],
 ['12224879',
  'Interactive Dynamic Video',
  'http://www.interactivedynamicvideo.com/',
  '386',
  '52',
  'ne0phyte',
  '8/4/2016 11:52'],
 ['10975351',
  'How to Use Open Source and Shut the Fuck Up at the Same Time',
  'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/',
  '39',
  '10',
  'josep2',
  '1/26/2016 19:30'],
 ['11964716',
  "Florida DJs May Face Felony for April Fools' Water Joke",
  'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/',
  '2',
  '1',
  'vezycash',
  '6/23/2016 22:20'],
 ['11919867',
  'Technology ventures: From Idea to Enterprise',
  'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429',
  '3',
  '1',
  'hswarna',
  '6/17/2016 0:01']]

In [4]:
# remove headers
headers=hn[0]
hn=hn[1:]
print(headers)
print(hn[:5])

['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at']
[['12224879', 'Interactive Dynamic Video', 'http://www.interactivedynamicvideo.com/', '386', '52', 'ne0phyte', '8/4/2016 11:52'], ['10975351', 'How to Use Open Source and Shut the Fuck Up at the Same Time', 'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/', '39', '10', 'josep2', '1/26/2016 19:30'], ['11964716', "Florida DJs May Face Felony for April Fools' Water Joke", 'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/', '2', '1', 'vezycash', '6/23/2016 22:20'], ['11919867', 'Technology ventures: From Idea to Enterprise', 'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429', '3', '1', 'hswarna', '6/17/2016 0:01'], ['10301696', 'Note by Note: The Making of Steinway L1037 (2007)', 'http://www.nytimes.com/2007/11/07/movies/07stein.html?_r=0', '8', '2', 'walterbell', '9/30/2015 4:12']]


We can see above that the data set contains the title of the posts, the number of comments for each post, and the date the post was created. 

## Extracting Ask HN and Show HN Posts

Since we're only concerned with post titles beginning with Ask HN or Show HN, we'll create new lists of lists containing just the data for those titles.

In [5]:
# Separating Data
ask_posts=[]
show_posts=[]
other_posts=[]
for row in hn:
    title=row[1].lower()
    if title.startswith('ask hn'):
        ask_posts.append(row)
    elif title.startswith('show hn'):
        show_posts.append(row)
    else:
        other_posts.append(row)
        
print(len(ask_posts))
print(len(show_posts))
print(len(other_posts))

1744
1162
17194


We can see there are 1744 Ask HN posts, 1162 Show HN posts, and 17194 other posts.

## Calculating the Average Number of Comments for Ask HN and Show HN Posts

We have separated the "ask posts" and the "show posts" into two list of lists named ask_posts and show_posts. Next, let's determine if ask posts or show posts receive more comments on average.

In [21]:
# Calculate the average number of comments of ask posts
total_ask_comments=0
for row in ask_posts:
    num_comments=int(row[4])
    total_ask_comments+=num_comments
avg_ask_comments=total_ask_comments/len(ask_posts)
print(avg_ask_comments)

14.038417431192661


In [22]:
# Calculate the average number of comments of show posts
total_show_comments=0
for row in show_posts:
    num_comments=int(row[4])
    total_show_comments+=num_comments
avg_show_comments=total_show_comments/len(show_posts)
print(avg_show_comments)

10.31669535283993


We can see that ask posts receive more comments than show posts. Since ask posts are more likely to receive comments, we'll focus our remaining analysis just on these posts.

## Finding the Amount of Ask Posts and Comments by Hour Created

In [ ]:
# Calculate the amount of ask posts created during each hour of day and the number of comments received.
import datetime as dt

result_list=[]
for row in ask_posts:
    result_list.append([row[6], int(row[4])])
    
counts_by_hour={}
comments_by_hour={}
for row in result_list:
    date=row[0]
    date=dt.datetime.strptime(date, "%m/%d/%Y %H:%M")
    hour=dt.datetime.strftime(date, "%H")
    if hour not in counts_by_hour:
        counts_by_hour[hour]=1
        comments_by_hour[hour]=row[1]
    else:
        counts_by_hour[hour]+=1
        comments_by_hour[hour]+=row[1]

In [23]:
print(comments_by_hour)

{'12': 687, '19': 1188, '13': 1253, '10': 793, '09': 251, '20': 1722, '21': 1745, '06': 397, '04': 337, '08': 492, '16': 1814, '14': 1416, '01': 683, '02': 1381, '07': 267, '18': 1439, '17': 1146, '03': 421, '05': 464, '11': 641, '00': 447, '22': 479, '15': 4477, '23': 543}


## Calculating the Average Number of Comments for Ask HN Posts by Hour

In [15]:
# Calculate the average amount of comments `Ask HN` posts created at each hour of the day receive.
avg_by_hour=[]
for hour in comments_by_hour:
    avg_by_hour.append([hour, comments_by_hour[hour]/counts_by_hour[hour]])
avg_by_hour

[['12', 9.41095890410959],
 ['19', 10.8],
 ['13', 14.741176470588234],
 ['10', 13.440677966101696],
 ['09', 5.5777777777777775],
 ['20', 21.525],
 ['21', 16.009174311926607],
 ['06', 9.022727272727273],
 ['04', 7.170212765957447],
 ['08', 10.25],
 ['16', 16.796296296296298],
 ['14', 13.233644859813085],
 ['01', 11.383333333333333],
 ['02', 23.810344827586206],
 ['07', 7.852941176470588],
 ['18', 13.20183486238532],
 ['17', 11.46],
 ['03', 7.796296296296297],
 ['05', 10.08695652173913],
 ['11', 11.051724137931034],
 ['00', 8.127272727272727],
 ['22', 6.746478873239437],
 ['15', 38.5948275862069],
 ['23', 7.985294117647059]]

## Sorting and Printing Values from a List of Lists

We have calculated the average number of comments for posts created during each hour of the day, and stored the results in a list of lists named avg_by_hour.

Although we now have the results we need, this format makes it hard to identify the hours with the highest values. Let's finish by sorting the list of lists and printing the five highest values in a format that's easier to read.

In [26]:
swap_avg_by_hour = []

for row in avg_by_hour:
    swap_avg_by_hour.append([row[1],row[0]])
    
print(swap_avg_by_hour)

[[9.41095890410959, '12'], [10.8, '19'], [14.741176470588234, '13'], [13.440677966101696, '10'], [5.5777777777777775, '09'], [21.525, '20'], [16.009174311926607, '21'], [9.022727272727273, '06'], [7.170212765957447, '04'], [10.25, '08'], [16.796296296296298, '16'], [13.233644859813085, '14'], [11.383333333333333, '01'], [23.810344827586206, '02'], [7.852941176470588, '07'], [13.20183486238532, '18'], [11.46, '17'], [7.796296296296297, '03'], [10.08695652173913, '05'], [11.051724137931034, '11'], [8.127272727272727, '00'], [6.746478873239437, '22'], [38.5948275862069, '15'], [7.985294117647059, '23']]


In [27]:
sorted_swap = sorted(swap_avg_by_hour, reverse=True)
sorted_swap

[[38.5948275862069, '15'],
 [23.810344827586206, '02'],
 [21.525, '20'],
 [16.796296296296298, '16'],
 [16.009174311926607, '21'],
 [14.741176470588234, '13'],
 [13.440677966101696, '10'],
 [13.233644859813085, '14'],
 [13.20183486238532, '18'],
 [11.46, '17'],
 [11.383333333333333, '01'],
 [11.051724137931034, '11'],
 [10.8, '19'],
 [10.25, '08'],
 [10.08695652173913, '05'],
 [9.41095890410959, '12'],
 [9.022727272727273, '06'],
 [8.127272727272727, '00'],
 [7.985294117647059, '23'],
 [7.852941176470588, '07'],
 [7.796296296296297, '03'],
 [7.170212765957447, '04'],
 [6.746478873239437, '22'],
 [5.5777777777777775, '09']]

In [28]:
# Print the the 5 hours with the highest average comments.
print("Top 5 Hours for Ask Posts Comments")
for avg,hr in sorted_swap[:5]:
    print("{}:{:.2f} average comments per post".format(dt.datetime.strptime(hr,"%H").strftime("%H:%M"), avg))

Top 5 Hours for Ask Posts Comments
15:00:38.59 average comments per post
02:00:23.81 average comments per post
20:00:21.52 average comments per post
16:00:16.80 average comments per post
21:00:16.01 average comments per post


The hour that receives the most comments per post on average is 15:00, with an average of 38.59 comments per post. There's about a 60% increase in the number of comments between the hours with the highest and second highest average number of comments.

## Conclusion

In this project, we analyzed ask posts and show posts to determine which type of post and time receive the most comments on average. Based on our analysis, to maximize the amount of comments a post receives, we'd recommend the post be categorized as ask post and created between 15:00 and 16:00.

However, it should be noted that the data set we analyzed excluded posts without any comments. Given that, it's more accurate to say ask posts received more comments on average and ask posts created between 15:00 and 16:00 (3:00 pm est - 4:00 pm est) received the most comments on average.